scipy/spatial/ckdtree/src/build.cxx

```diff
@@ -142,62 +127,259 @@ build(ckdtree *self, ckdtree_intp_t start_idx, intptr_t end_idx,
+
+        const double split = node->split;
+        const ckdtree_intp_t d = node->split_dim;
+
         if (CKDTREE_LIKELY(_compact)) {
-            _less = build(self, start_idx, p, maxes, mins, _median, _compact);
-            _greater = build(self, p, end_idx, maxes, mins, _median, _compact);
+            _less = build(self, start_idx, p, maxes, mins, _median, _compact, buf);
+            _greater = build(self, p, end_idx, maxes, mins, _median, _compact, buf);
         }
         else
         {
-            std::vector<double> tmp(m);
-            double *mids = &tmp[0];
+            std::vector<double> mids(self->m);
 
-            for (i=0; i<m; ++i) mids[i] = maxes[i];
+            std::copy_n(maxes, mids.size(), mids.begin());
             mids[d] = split;
-            _less = build(self, start_idx, p, mids, mins, _median, _compact);
+            _less = build(self, start_idx, p, mids.data(), mins, _median, _compact, buf);
 
-            for (i=0; i<m; ++i) mids[i] = mins[i];
+            std::copy_n(mins, mids.size(), mids.begin());
             mids[d] = split;
-            _greater = build(self, p, end_idx, maxes, mids, _median, _compact);
+            _greater = build(self, p, end_idx, maxes, mids.data(), _median, _compact, buf);
         }
 
-        /* recompute n because std::vector can
-         * reallocate its internal buffer
-         */
-        root = tree_buffer_root(self->tree_buffer);
-        n = root + node_index;
+        node = &buf[node_index];
         /* fill in entries */
-        n->_less = _less;
-        n->_greater = _greater;
-        n->less = root + _less;
-        n->greater = root + _greater;
-        n->split_dim = d;
-        n->split = split;
+        node->_less = _less;
+        node->less = buf.data() + node->_less;
+        node->_greater = _greater;
+        node->greater = buf.data() + node->_greater;
 
         return node_index;
     }
 }
 
+static ckdtree_intp_t
+build(ckdtree *self, ckdtree_intp_t start_idx, intptr_t end_idx,
+    double *maxes, double *mins,
+    const int _median, const int _compact,
+    std::list<std::vector<ckdtreenode>>* buflist, int workers);
+
+struct build_worker {
+
+    std::list<std::vector<ckdtreenode>> buflist_;
+    std::vector<double> mins_, maxes_;
+    std::future<ckdtree_intp_t> future_;
+
+    ckdtree_intp_t run(ckdtree *self, ckdtree_intp_t start_idx, intptr_t end_idx,
+          const double *maxes, const double *mins,
+          const int _median, const int _compact,
+          int workers)
+    {
+        const int subworkers = workers/2;
+        assert(subworkers > 0);
+
+        mins_.resize(self->m);
+        std::copy_n(mins, mins_.size(), mins_.begin());
+        maxes_.resize(self->m);
+        std::copy_n(maxes, maxes_.size(), maxes_.begin());
+
+        future_ = std::async(std::launch::async, [&]() {
+                buflist_.emplace_back();
+                buflist_.back().reserve((end_idx-start_idx)/self->leafsize/workers);
```

Comment:
Why call `emplace_back` and `reserve` each time you call build, instead of just calling it inside of `build`?

Reply:
In a way this is left over from when there was only one build for both the parallel and the sequential recursion. Now it basically does not matter, both places are equivalent in terms of code sequence. However, as `buflist_` is a member of `build_worker`, so it seems sensible to me to initialized it here.

Reply:
> Now it basically does not matter, both places are equivalent in terms of code sequence. 

It is equivalent yes, but this way it's copied to two places that must be kept in sync vs. only one place at the top of `build`.

Reply:
Fixed in https://github.com/scipy/scipy/pull/14481/commits/1776bb188f00738d69f0f727f1ee54eac9c86806 .